File: Mika/tags/PKG-adhoco-4/core-vm/java/java/util/zip/Adler32.java
/**************************************************************************
* Copyright (c) 2001 by Punch Telematix. All rights reserved.             *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of Punch Telematix nor the names of                 *
*    other contributors may be used to endorse or promote products        *
*    derived from this software without specific prior written permission.*
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL PUNCH TELEMATIX OR OTHER CONTRIBUTORS BE LIABLE       *
* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR            *
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF    *
* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR         *
* BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,   *
* WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE    *
* OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN  *
* IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.                           *
**************************************************************************/

/*
** $Id: Adler32.java,v 1.2 2006/02/17 10:53:19 cvs Exp $
*/

package java.util.zip;

/**
*         Adler-32 is composed of two sums accumulated per byte: s1 is
*         the sum of all bytes, s2 is the sum of all s1 values. Both sums
*         are done modulo 65521. s1 is initialized to 1, s2 to zero.  The
*         Adler-32 checksum is stored as s2*65536 + s1 in most-
*         significant-byte first (network) order.
*/	
public class Adler32 implements Checksum {

	private int s1=1;
	private int s2;
	
	
	public Adler32() { }
	
	public long getValue(){
	 	return s2 * 65536L + s1;
	}
	
	public void reset(){
		s1=1;
		s2=0;
	}
	
	public void update(int bval) {
		s1 = ((0x0FF & bval) + s1) % 65521;
		s2 = (s2 + s1) % 65521;	
		
	}
	public void update(byte[] buf) {
	 	update(buf, 0, buf.length);
	}
	/**
	* this method is not that hard to do in java code, but it is very likely that large byte arrays are passed
	* to this method --> so it can be justified to get the job done in native code ...
	*/
	public native void update(byte[] buf, int off, int len);

}

File: core-vm/src/native/java/util/zip/Adler32.c
/**************************************************************************
* Copyright (c) 2021 by KIFFER Ltd. All rights reserved.                  *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of KIFFER Ltd nor the names of other contributors   *
*    may be used to endorse or promote products derived from this         *
*    software without specific prior written permission.                  *
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL KIFFER LTD OR OTHER CONTRIBUTORS BE LIABLE FOR ANY    *
* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL      *
* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE       *
* GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS           *
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER    *
* IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR         *
* OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF  *
* ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.                              *
**************************************************************************/

#include <string.h>

#include "clazz.h"
#include "arrays.h"
#include "exception.h"
#include "heap.h"
#include "core-classes.h"


void Adler32_update(w_thread thread, w_instance thisAdler32, w_instance byteArray, w_int off, w_int len) {
  w_int sum1 = getIntegerField(thisAdler32, F_Adler32_s1);
  w_int sum2 = getIntegerField(thisAdler32, F_Adler32_s2);
  
    if (!byteArray) {
          throwException(thread, clazzNullPointerException, NULL);
    }
  else {
      if (off < 0 || len < 0 || off > instance2Array_length(byteArray) - len) {
          throwException(thread, clazzArrayIndexOutOfBoundsException, NULL);
      }
          else {
            w_ubyte * data = (w_ubyte *) instance2Array_byte(byteArray);
            w_int loop=0;
            w_int count;  
            
            data += off;
      /**
      ** we want to avoid to do the modulo 65521 after every sum ...
      ** worst case scenario s1 == 65520 and all new bytes are 0xff == 255
      ** max int = 2147483647
      ** we should calculate the modulo thing once every 8000000 times for s1
      ** s2 = s1 + s2 , with s1 = 65520 + n * 255
      ** ==> s2 = (65520 + n * 255) + (65520 + (n-1) * 255 ) + ... + 65520
      ** ==> s2 = (n+1) * 65520 + 255 * (n^2 + n)
      ** ==> 2147483647 = 255 * n^2 + 65775 * n + 65520
      **  n^2 + 258 * n  - 8400000 = 0
      ** --> nmax ~ 2500
      ** ==> to be on the safe side we take n = 2000
      */
               while ( len > 0 ) {
                  
                  loop = (len > 2000 ? 2000 : len);
                  for (count=0 ; count < loop ; count++) {
                     sum1 += *(data++);
                     sum2 += sum1;
                  }
                  sum1 = sum1 % 65521;
                  sum2 = sum2 % 65521;
                  len -= loop;
               }
               setIntegerField(thisAdler32, F_Adler32_s1, sum1);
         setIntegerField(thisAdler32, F_Adler32_s2, sum2);
         }
  }  

}

